src/python/grpcio_observability/grpc_observability/observability_util.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(xuanwn): clean up includes
+#include <stdint.h>
+
+#include <algorithm>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/status.h>
+
+#include "src/python/grpcio_observability/grpc_observability/constants.h"
+#include "src/python/grpcio_observability/grpc_observability/python_census_context.h"
+
+#ifndef OBSERVABILITY_MAIN_H
+#define OBSERVABILITY_MAIN_H
+
+namespace grpc_observability {
+
+struct CensusData {
+  DataType type;
+  std::vector<Label> labels;
+  // TODO(xuanwn): We can use union here
+  SpanCensusData span_data;
+  Measurement measurement_data;
+  CensusData() {}
+  CensusData(const Measurement& mm, std::vector<Label>& labels)
+      : type(kMetricData), labels(std::move(labels)), measurement_data(mm) {}
+  CensusData(const SpanCensusData& sd) : type(kSpanData), span_data(sd) {}
+};
+
+struct CloudMonitoring {
+  CloudMonitoring() {}
+};
+
+struct CloudTrace {
+  float sampling_rate = 0.0;
+  CloudTrace() {}
+  CloudTrace(double sr) : sampling_rate(sr) {}
+};
+
+struct CloudLogging {
+  CloudLogging() {}
+};
+
+struct GcpObservabilityConfig {
+  CloudMonitoring cloud_monitoring;
+  CloudTrace cloud_trace;
+  CloudLogging cloud_logging;
+  std::string project_id;
+  std::vector<Label> labels;
+  bool is_valid;
+  GcpObservabilityConfig() : is_valid(false) {}
+  GcpObservabilityConfig(bool valid) : is_valid(true) {}
+  GcpObservabilityConfig(CloudMonitoring cloud_monitoring,
+                         CloudTrace cloud_trace, CloudLogging cloud_logging,
+                         std::string project_id, std::vector<Label> labels)
```

Comment:
`const std::string&`
`const std::vector<Label>&`

In general, any type that is large or may be backed by heap allocations should be passed by const ref. This includes `std::string`, `std::vector`, and all other containers. It also includes all classes/structs that include these as fields. Please do a a scrub through the PR for this.